library/src/main/scala/org/clulab/processors/clu/BalaurProcessor.scala

```diff
@@ -34,230 +35,248 @@ class BalaurProcessor protected (
+          sentence
+        case e: AssertionError =>
+          println(s"ERROR: The output of predictWithScores does not satisfy assertions.  The sentence will not be annotated: ${sentence.words.mkString(" ")}")
+          sentence
       }
     }
+    val partlyAnnotatedDocument = doc.copy(sentences = partlyAnnotatedSentences)
+    val fullyAnnotatedDocument = numericEntityRecognizerOpt.map { numericEntityRecognizer =>
+      val numericMentions = numericEntityRecognizer.extractFrom(partlyAnnotatedDocument)
+      val (newLabels, newNorms) = NumericUtils.mkLabelsAndNorms(partlyAnnotatedDocument, numericMentions)
+      val fullyAnnotatedSentences = partlyAnnotatedDocument.sentences.indices.map { index =>
+        partlyAnnotatedDocument.sentences(index).copy(
+          entities = Some(newLabels(index)),
+          norms = Some(newNorms(index))
+        )
+      }
 
-    // numeric entities using our numeric entity recognizer based on Odin rules
-    if(numericEntityRecognizerOpt.nonEmpty) {
-      val numericMentions = extractNumericEntityMentions(doc)
-      setLabelsAndNorms(doc, numericMentions)
-    }
+      partlyAnnotatedDocument.copy(sentences = fullyAnnotatedSentences)
+    }.getOrElse(partlyAnnotatedDocument)
 
-    doc
+    fullyAnnotatedDocument
   }
 
-  def extractNumericEntityMentions(doc:Document): Seq[Mention] = {
-    numericEntityRecognizerOpt.get.extractFrom(doc)
-  }
+  private def mkPosTags(words: Seq[String], labels: Array[Array[(String, Float)]]): Seq[String] = {
+    assert(labels.length == words.length)
 
-  private def assignPosTags(labels: Array[Array[(String, Float)]], sent: Sentence): Unit = {
-    assert(labels.length == sent.words.length)
-    sent.tags = Some(postprocessPartOfSpeechTags(sent.words, labels.map(_.head._1).toArray))
-  }
+    val rawTags = WrappedArraySeq(labels.map(_.head._1)).toImmutableSeq
+    val cookedTags = PostProcessor.postprocessPartOfSpeechTags(words, rawTags)
 
-  /** Must be called after assignPosTags and lemmatize because it requires Sentence.tags and Sentence.lemmas */
-  private def assignNamedEntityLabels(labels: Array[Array[(String, Float)]], sent: Sentence): Unit = {
-    assert(labels.length == sent.words.length)
+    cookedTags
+  }
 
-    // NER labels from the custom NER
-    val optionalNERLabels: Option[Array[String]] = optionalNER.map { ner =>
+  private def mkNerLabelsOpt(
+    words: Seq[String], startOffsets: Seq[Int], endOffsets: Seq[Int],
+    tags: Seq[String], lemmas: Seq[String]
+  ): Option[Seq[String]] = {
+    lexiconNerOpt.map { lexiconNer =>
       val sentence = Sentence(
-        sent.words,
-        sent.startOffsets,
-        sent.endOffsets,
-        sent.words,
-        sent.tags,
-        sent.lemmas,
-        entities = None,
-        norms = None,
-        chunks = None,
-        tree = None,
-        deps = EMPTY_GRAPH,
-        relations = None
+        words, // TODO: Why isn't this raw?
```

Comment:
Here is a worrisome TODO.

Reply:
I vaguely remember this... I think some lexicon NER dictionaries operate over lemmas rather than words, which required such weird work arounds...